File: darwin/flutter_compress_pro/Sources/flutter_compress_pro/PluginFiles.swift
import Foundation

/// The plugin's private scratch directory for intermediate outputs/thumbnails.
enum PluginFiles {
  static func cacheDir() -> URL {
    let dir = FileManager.default.temporaryDirectory.appendingPathComponent("flutter_compress_pro")
    try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    return dir
  }

  static func clearCache() {
    let items = (try? FileManager.default.contentsOfDirectory(
      at: cacheDir(), includingPropertiesForKeys: nil)) ?? []
    for item in items { try? FileManager.default.removeItem(at: item) }
  }

  /// Compose the output URL. The name is `outputName` (extension stripped — the
  /// caller's contract is "base name only") or the source's base name plus a
  /// timestamp; `ext` (without dot) is always appended so it matches the bytes
  /// written. Writes into `outputDir` when it's a real writable directory,
  /// otherwise the plugin cache.
  static func resolveOutput(
    outputDir: String?, outputName: String?, sourcePath: String, ext: String
  ) -> URL {
    let base: String
    if let name = outputName, !name.isEmpty {
      base = ((name as NSString).lastPathComponent as NSString).deletingPathExtension
    } else {
      let srcBase = ((sourcePath as NSString).lastPathComponent as NSString).deletingPathExtension
      base = "\(srcBase)_\(Int(Date().timeIntervalSince1970 * 1000))"
    }
    let fileName = "\(base).\(ext)"
    var target = cacheDir().appendingPathComponent(fileName)
    if let dir = outputDir, dir.hasPrefix("/") {
      let dirURL = URL(fileURLWithPath: dir, isDirectory: true)
      try? FileManager.default.createDirectory(at: dirURL, withIntermediateDirectories: true)
      if FileManager.default.isWritableFile(atPath: dir) {
        target = dirURL.appendingPathComponent(fileName)
      }
    }
    // Never write over the input: the engines still read the source afterwards
    // (EXIF copy, size comparison), and clobbering it in place loses the original.
    let sourceURL = URL(fileURLWithPath: sourcePath)
    if target.standardizedFileURL.path == sourceURL.standardizedFileURL.path {
      return target.deletingLastPathComponent().appendingPathComponent("\(base)-1.\(ext)")
    }
    return target
  }
}

File: android/src/main/kotlin/com/compress/all/flutter_compress_pro/PluginFiles.kt
package com.compress.all.flutter_compress

import android.content.Context
import java.io.File

/** The plugin's private scratch directory for intermediate outputs/thumbnails. */
internal fun Context.compressCacheDir(): File =
    File(cacheDir, "flutter_compress").apply { mkdirs() }

internal fun Context.clearCompressCache() {
    compressCacheDir().listFiles()?.forEach { it.delete() }
}

/**
 * Compose the output file. The name is [outputName] (extension stripped — the
 * caller's contract is "base name only") or, when absent, the source's base
 * name plus a timestamp. [ext] (without dot) is always appended so the
 * extension matches the bytes actually written. Writes into [outputDir] when
 * it's a real writable path, otherwise the plugin cache.
 */
internal fun Context.resolveOutput(
    outputDir: String?,
    outputName: String?,
    sourcePath: String,
    ext: String,
): File {
    val base = outputName
        ?.substringAfterLast('/')?.substringAfterLast('\\')
        // Drop any extension the caller supplied — [ext] is authoritative, so
        // "photo.jpg" must not become "photo.jpg.jpg".
        ?.substringBeforeLast('.')
        ?.takeIf { it.isNotBlank() }
        ?: "${File(sourcePath).nameWithoutExtension}_${System.currentTimeMillis()}"
    val fileName = "$base.$ext"
    val target = if (outputDir != null && outputDir.startsWith("/")) {
        val dir = File(outputDir).also { it.mkdirs() }
        if (dir.canWrite()) File(dir, fileName) else File(compressCacheDir(), fileName)
    } else {
        File(compressCacheDir(), fileName)
    }
    // Never write over the input: the engines still read the source afterwards
    // (EXIF copy, size comparison), and clobbering it in place loses the original.
    if (target.absolutePath == File(sourcePath).absolutePath) {
        return File(target.parentFile, "$base-1.$ext")
    }
    return target
}
